inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp

```diff
@@ -667,62 +679,79 @@ void RemovePermutationsNHWCToNCHWPass::run() {
             }
 
             if (getPassManager()->getPolicy().NHWCToNCHWPolicy == Policy::NHWCToNCHW::REMOVE_ALL) {
-                permutationsToRemove.push_back(prev);
+                permutations_to_remove.insert(prev);
             }
         } else  {
-            if (prev->outData.size() != 1 || getInputTo(prev->outData[0]).size() != 1) {
+            if (parent->outData.size() != 1 || getInputTo(parent->outData[0]).size() != 1) {
                 continue;
             }
-            auto prev_dims = prev->outData[0]->getDims();
+            auto parent_dims = parent->outData[0]->getDims();
             // Check if the previous layer has all dimensions except one to be equal to 1
-            if (std::count_if(std::begin(prev_dims), std::end(prev_dims), [](size_t dim) { return dim != 1; }) > 1) {
+            if (std::count_if(std::begin(parent_dims), std::end(parent_dims), [](size_t dim) { return dim != 1; }) > 1) {
                 continue;
             }
         }
-        permutationsToRemove.push_back(next);
+        permutations_to_remove.insert(next);
+        nhwc_layout_patterns.push_back({prev, next});
+
+        auto* convolution = dynamic_cast<ConvolutionLayer*>(l.get());
+        if (!convolution) {
+            THROW_GNA_EXCEPTION << "There needs to be a convolution between permutations for RemovePermutationsNHWCToNCHWPass!";
```

Comment:
Should it be exception or you should skip to remove this permutations?

Reply:
I think it should be exception since LayerInfo(l).isConvolution() returned true if we are here but casting to ConvolutionLayer* failed.